### PR TITLE
added assertion for json body

### PR DESCRIPTION
**Pull request by katarinarosiak**
Co-authored-by: Tim Dronkers <tdronkers@gmail.com>
Co-authored-by: Scott Graham <scttgrhm7+public@gmail.com>
Co-authored-by: Miles Abbason <miles.abbason@gmail.com>

This PR: 
- adds case for `jsonBody` assertion type
- adds `utils/helpers.js` file for helper functions 

It assumes that:
- the value for `property` has a format `$.propName[0]` where `$` represents the response body and the remaining part represents the path to access the value in that JSON object.
    - Formats of `property` that are allowed: `propertName` and `propertyName.nestedKey[0]` (without `$` will automatically prepend `$.`) 
    - the value of `property` returned from the `testRunner` will always contain `$.` sign prepended. 
 - `contains` and `not_contains`comparison types:
     -  tested on objects or arrays will check if any key or value nested within the object contains a key or value equal to `target`. 
     - Tested on `undefined`, `null`, numbers will set `success` to `false`. 
     - Tested on boolean will check for equality.
 However, the case will not check whether the key or value's name contains the expected value. 
For example, the following JSON would not result in success = true if the `target` is `time`: 
```json
[
  {
   "key": [{}, {}, {"timetime": true}]
  }
]
```
- `has_key`, `not_has_key`, `has_value`, `not_has_value` will recursively search for keys or values within all objects and arrays in the response body. For example, if we want to check if body `has_key`: `time` the Test Runner checking the following JSON would set `success` to true:
```json
[
  {
   "key": [{}, {}, {"time": true}]
  }
]
```
However, the case will not check whether the key or value's name contains the expected value. 
For example, the following JSON would not result in success = true: 
```json
[
  {
   "key": [{}, {}, {"timetime": true}]
  }
]
```
- `less_than`, `greater_than`, `greater_than_or_equal_to`, `less_than_or_equal_to` are numbers and anything else would most likely result in an error so there should be some input validation. It does take into consideration that the `test` JSON is stringified along the way so it would coerce the input to a number. 
- `is_empty`, `is_not_empty` consider empty to be the following `{}`, `[]`, `""`  only. Thus, `null`, `undefined` or `0` are not considered to be empty.

### CODE REVIEW

File: entities/TestConfiguration.js [hunk 1]
```diff
@@ -45,27 +53,49 @@ class TestConfiguration {
...
             });
-          })
-        break;
+          });
+          break;
+        case 'jsonBody':
+          bodyAssertions = this.assertions[assertionType];
+          bodyAssertions.forEach((assertion) => {
+            targetValue = assertion.target || null;
+            property = assertion.property[0] !== '$' ? `$.${assertion.property}` : assertion.property;
```
Comment (scogra17) [inline entities/TestConfiguration.js:82]: Nice!

File: entities/TestConfiguration.js [hunk 2]
```diff
@@ -45,27 +53,49 @@ class TestConfiguration {
...
+            comparisonType = assertion.comparison;
+            actualValue = null;
+
+            if ((!response.data) || (!Array.isArray(response.data) && typeof response.data !== 'object')) {
+              success = false;
+            } else {
+              const responseBody = response.data;
+              actualValue = property ? helpers.getValue(responseBody, property) : responseBody;
```
Comment (scogra17) [inline entities/TestConfiguration.js:90]: Great call extracting this to a helper 👍

File: entities/TestConfiguration.js [hunk 3]
```diff
@@ -121,18 +151,114 @@ class TestConfiguration {
...
+        result = Number(responseHeaders[assertionHeader.property])
+        <= Number(assertionHeader.target);
+        break;
+      default:
+        result = false;
+    }
+    return result;
+  }
+
+  static checkJsonBody(targetValue, actualValue, comparisonType) {
```
Comment (scogra17) [inline entities/TestConfiguration.js:174]: Wondering if we could combine some of the `checkJsonBody` and `checkHeaders` code? Not crucial for v1 but maybe we should add as a TODO?

File: entities/utils/helpers.js [hunk 4]
```diff
@@ -0,0 +1,58 @@
+const containsKeysOrVals = (value, target) => {
```
Comment (scogra17) [inline entities/utils/helpers.js:1]: Another non-crucial comment for v1 -- these types of simple functions might be prime candidates for unit testing, which would serve not only to validate their functionality with a variety of inputs, but also as documentation

**Review by scogra17 — APPROVED** [review]
Nice work @katarinarosiak!! You added a lot of functionality here. I've added a couple comments related to keeping the code "DRY" and unit testing, but I don't think these are crucial at this stage

**Review by tdronkers04 — APPROVED** [review]
great work!

**Comment by scogra17** [comment]
P.S. Love the detailed description!

**Pull request merged**